community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java

```diff
@@ -126,4 +130,67 @@ void visit( ResourceType resourceType, long resourceId, String description, long
     void accept(Visitor visitor);
 
     void close();
+
+    abstract class ActiveLock
+    {
+        public final ResourceType resourceType;
+        public final long resourceId;
+
+        private ActiveLock( ResourceType resourceType, long resourceId )
+        {
+            this.resourceType = resourceType;
+            this.resourceId = resourceId;
+        }
+
+        @Override
+        public boolean equals( Object o )
+        {
+            if ( this == o )
+            {
+                return true;
+            }
+            if ( o.getClass() != this.getClass() )
+            {
+                return false;
+            }
+            ActiveLock activeLock = (ActiveLock) o;
+            return resourceId == activeLock.resourceId &&
+                    Objects.equals( resourceType, activeLock.resourceType );
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash( resourceType, resourceId );
+        }
+
+        public abstract String mode();
+    }
+
+    final class ActiveExclusiveLock extends ActiveLock
+    {
+        public ActiveExclusiveLock( ResourceType resourceType, long resourceId )
+        {
+            super( resourceType, resourceId );
+        }
+
+        @Override
+        public String mode()
+        {
+            return "EXCLUSIVE";
+        }
+    }
+    final class ActiveSharedLock extends ActiveLock
+    {
+        public ActiveSharedLock( ResourceType resourceType, long resourceId )
+        {
+            super( resourceType, resourceId );
+        }
+
+        @Override
+        public String mode()
+        {
+            return "SHARED";
+        }
+    }
```

Comment:
I guess you use this pattern to make the active locks collection use less heap memory?